components/uniset/src/uniset.rs

```diff
@@ -0,0 +1,387 @@
+            serialize
+        ))?
+        .parse()?;
+    if capacity % 2 != 0 {
+        return Err(format!("Capacity must be even. Parsed Capacity: {}", capacity).into());
+    }
+    let mut serialized_vec: Vec<u32> = Vec::with_capacity(capacity);
+    let mut prev: Option<u32> = None;
+    for str_elem in serialize {
+        let parsed: u32 = str_elem.parse()?;
+        if serialized_vec.len() + 1 > serialized_vec.capacity() {
+            return Err(format!(
+                "Serialization capacity is too small. Allocated Capacity: {}",
+                capacity
+            )
+            .into());
+        }
+        if Some(parsed) <= prev {
+            return Err(format!(
+                "Serialization must be sorted. {:?} followed by {:?}",
+                Some(parsed),
+                prev
+            )
+            .into());
+        }
+        serialized_vec.push(parsed);
+        prev = Some(parsed);
+    }
+    if serialized_vec.len() % 2 != 0 {
+        return Err(format!(
+            "Serialization must be even. Serialization: {:?} Length: {}",
+            serialized_vec,
+            serialized_vec.len()
+        )
+        .into());
+    }
+    serialized_vec.shrink_to_fit(); // necessary if the length < capacity
+    Ok(serialized_vec)
+}
+
+/// UnicodeSet membership wrapper
+///
+/// Provides exposure to membership functions and constructors from serialized UnicodeSets
+/// and predefined ranges.
+/// Implements an inversion list.
+//#[derive(Copy, Clone, Debug, Eq)]
+pub struct UnicodeSet {
+    // If we wanted to use an array to keep the memory on the stack, there is an unsafe nightly feature
+    // https://doc.rust-lang.org/nightly/core/array/trait.FixedSizeArray.html
+    // Allows for traits of fixed size arrays
+    inv_list: Vec<u32>,
+}
+
+impl FromStr for UnicodeSet {
+    type Err = Box<dyn Error>;
+
+    fn from_str(serialize: &str) -> Result<Self, Self::Err> {
+        match parse_serial_string(serialize) {
+            Ok(serialize) => {
+                if serialize.len() % 2 != 0 {
+                    return Err("Array length must be even".into());
+                }
+                Ok(UnicodeSet {
+                    inv_list: serialize,
+                })
+            }
+            Err(e) => Err(e.into()),
+        }
+    }
+}
+
+impl UnicodeSet {
+    /// Returns Result of UnicodeSet from serialized string
+    ///
+    /// Returns an error if the serialized string fails to parse.
+    /// The serialized string requires starting capacity integer, followed by space delimited
+    /// integer code points. There must be an even number of elements (not including the
+    /// capacity int), and must be in ascending sorted order.
+    ///
+    /// Example String: `"4 0 5 10 15"` designates a capacity of size `4`, followed by 2 ranges
+    /// The ranges are `{0, 4}` and `{10, 14}` inclusive
+    pub fn new(serialize: &str) -> Result<UnicodeSet, Box<dyn Error>> {
```

Comment:
I don't think we should have `new` at all. `new` is a constructor for cases where you have some default value that you initialize with some options. If you're transforming one input into another, esp. in a fallible way, use `From` (like `FromStr`).

If you'll go after Shane's suggestion to not use source string and serialize it, then you can do:

```rust
From<Vec<Range<u32>>> for UnicodeSet;
```

with some debug assertions about the input being sorted etc. If you want to allow for input to be not sorted, use `TryFrom`.

Reply:
Due to the memory and performance benefits we are sticking with `Vec<u32>` and performing the operations on that over the `Vec<Range<u32>>`. Why would we want to use `Box<[u32]>` over `Vec<u32>`? Since the size of `inv_list` doesn't change upon initialization from the `try_from`, a `Box` seems rather redundant.

Reply:
agreed and changed regarding the removal of `new`